File: go/pkg/memreaderwriter.go
package pkg

import "bytes"

// MemReaderWriter is an in-memory implementation of ChunkWriter and ByteAndBlockReader interfaces
// that allows to first write to the buffer and then read from it.
// Typically used in tests.
type MemReaderWriter struct {
	buf bytes.Buffer
}

func (m *MemReaderWriter) ReadByte() (byte, error) {
	return m.buf.ReadByte()
}

func (m *MemReaderWriter) Read(p []byte) (n int, err error) {
	return m.buf.Read(p)
}

func (m *MemReaderWriter) WriteChunk(header []byte, content []byte) error {
	if _, err := m.buf.Write(header); err != nil {
		return err
	}
	_, err := m.buf.Write(content)
	return err
}

func (m *MemReaderWriter) Bytes() []byte {
	return m.buf.Bytes()
}

File: rust/stef-core/src/memreaderwriter.rs
use std::io::{Cursor, Read};

use crate::{chunkwriter::ChunkWriter, errors::Result};

/// In-memory duplex object for tests: write chunks, then read bytes.
#[derive(Debug, Default, Clone)]
pub struct MemReaderWriter {
    cursor: Cursor<Vec<u8>>,
}

impl MemReaderWriter {
    /// Returns full buffer bytes.
    pub fn bytes(&self) -> &[u8] {
        self.cursor.get_ref()
    }

    /// Appends raw bytes to backing buffer.
    pub fn append_bytes(&mut self, data: &[u8]) {
        self.cursor.get_mut().extend_from_slice(data);
    }

    /// Sets read position.
    pub fn set_position(&mut self, pos: u64) {
        self.cursor.set_position(pos);
    }

    /// Reads one byte.
    pub fn read_byte(&mut self) -> std::io::Result<u8> {
        let mut b = [0u8; 1];
        let n = self.cursor.read(&mut b)?;
        if n == 0 {
            return Err(std::io::Error::new(std::io::ErrorKind::UnexpectedEof, "EOF"));
        }
        Ok(b[0])
    }
}

impl Read for MemReaderWriter {
    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
        self.cursor.read(buf)
    }
}

impl ChunkWriter for MemReaderWriter {
    fn write_chunk(&mut self, header: &[u8], content: &[u8]) -> Result<()> {
        self.cursor.get_mut().extend_from_slice(header);
        self.cursor.get_mut().extend_from_slice(content);
        Ok(())
    }
}
